File: csharp/Frugal/Policy.cs
namespace Frugal;

public sealed class PolicyOptions
{
    /// <summary>Model the client would have used without frugal; defaults to the priciest frontier model.</summary>
    public string? BaselineId { get; set; }
    public double WeightDifficulty { get; set; } = 0.55;
    public double WeightReasoning { get; set; } = 0.30;
    public double WeightStakes { get; set; } = 0.15;
    /// <summary>Need score at which tier 2 starts.</summary>
    public double StandardThreshold { get; set; } = 0.28;
    /// <summary>Need score at which tier 3 starts.</summary>
    public double FrontierThreshold { get; set; } = 0.58;
    /// <summary>Below this difficulty confidence, bump one tier up for safety.</summary>
    public double MinConfidence { get; set; } = 0.45;
    /// <summary>Stakes probability above which economy models are excluded.</summary>
    public double HighStakes { get; set; } = 0.70;
}

/// <summary>
/// Pure policy: judgments + features + catalog -> decision. No I/O, so it is cheap to test and
/// cheap to re-run with new weights. Mirrors ts/src/policy.ts.
/// </summary>
public static class Policy
{
    /// <summary>Expected visible output tokens by output-size level (0, 1, 2).</summary>
    private static readonly int[] OutputTokens = [250, 1200, 5000];

    /// <summary>Output multiplier for reasoning tokens at each effort level.</summary>
    private static readonly Dictionary<string, double> EffortMultiplier = new()
    {
        ["none"] = 1, ["minimal"] = 1, ["low"] = 1.4, ["medium"] = 2.2, ["high"] = 3.5, ["xhigh"] = 5, ["max"] = 7,
    };

    public static Decision Decide(Features f, Judgment j, IReadOnlyList<ModelSpec> available, PolicyOptions? opts = null)
    {
        var o = opts ?? new PolicyOptions();
        var rationale = new List<string>();
        var task = j.Task;
        var d = j.Difficulty.Score;
        var r = j.Reasoning.Score;
        var s = j.Stakes;

        // 1. How much capability does this request need?
        var need = o.WeightDifficulty * (d / 4) + o.WeightReasoning * (r / 3) + o.WeightStakes * s;
        var tier = need < o.StandardThreshold ? 1 : need < o.FrontierThreshold ? 2 : 3;
        rationale.Add($"task={task} difficulty={d:F2} reasoning={r:F2} stakes={s:F2} need={need:F2} -> tier {tier}");

        if (j.Source == "fallback") rationale.Add($"jev unavailable ({j.Error ?? "unknown"}); using conservative defaults");
        if (s >= o.HighStakes && tier == 1)
        {
            tier = 2;
            rationale.Add($"stakes {s:F2} >= {o.HighStakes}: economy tier excluded");
        }
        if (j.Source == "jev" && j.Difficulty.Confidence < o.MinConfidence && tier < 3)
        {
            tier++;
            rationale.Add($"difficulty confidence {j.Difficulty.Confidence:F2} < {o.MinConfidence}: bumped to tier {tier}");
        }

        // 2. How hard should the chosen model think?
        var wantEffort = DesiredEffort(d, r, s);
        rationale.Add($"desired effort {wantEffort}");

        // 3. Filter candidates on hard requirements, then rank by expected cost.
        var outTokens = ExpectedOutputTokens(j.OutputSize.Score, f.RequestedMaxTokens);
        var inTokens = f.InputTokens;
        var scored = available
            .Where(m => MeetsRequirements(m, f, inTokens, outTokens))
            .Select(m => new Candidate(m.Id, EstimateCost(m, inTokens, outTokens, SnapEffort(wantEffort, m.Effort)), EffectiveTier(m, task)))
            .OrderBy(c => c.EstimatedCost)
            .ToList();

        var pool = scored.Where(c => c.EffectiveTier >= tier).ToList();
        if (pool.Count == 0)
        {
            var best = scored.Count > 0 ? scored.Max(c => c.EffectiveTier) : 0;
            pool = scored.Where(c => c.EffectiveTier == best).ToList();
            rationale.Add($"no eligible model at tier {tier}; using best available (tier {best})");
        }
        if (pool.Count == 0)
            throw new InvalidOperationException("frugal: no model satisfies the request (check provider keys, vision/tool support, context size)");

        var winner = pool[0];
        var chosen = available.First(m => m.Id == winner.Id);
        var effort = SnapEffort(wantEffort, chosen.Effort);
        if (EffectiveTier(chosen, task) > chosen.Tier) rationale.Add($"{chosen.Id} counts a tier higher for {task}");
        rationale.Add($"cheapest eligible: {chosen.Id}{(effort is null ? "" : $" @ {effort}")} (~${winner.EstimatedCost:F5})");

        // 4. What would the baseline have cost?
        var baseline = PickBaseline(available, o.BaselineId);
        var baselineCost = baseline is null
            ? winner.EstimatedCost
            : EstimateCost(baseline, inTokens, outTokens, SnapEffort("high", baseline.Effort));

        return new Decision
        {
            Model = chosen,
            Effort = effort,
            Tier = tier,
            Task = task,
            Judgment = j,
            Estimate = new Estimate
            {
                InputTokens = inTokens,
                OutputTokens = outTokens,
                Cost = winner.EstimatedCost,
                BaselineModel = baseline?.Id ?? chosen.Id,
                BaselineCost = baselineCost,
                Savings = Math.Max(0, baselineCost - winner.EstimatedCost),
            },
            Baseline = baseline,
            Rationale = rationale,
            Candidates = scored,
            Passthrough = false,
        };
    }

    public static string DesiredEffort(double difficulty, double reasoning, double stakes)
    {
        var e = reasoning < 0.5 ? "minimal" : reasoning < 1.25 ? "low" : reasoning < 2.0 ? "medium" : reasoning < 2.6 ? "high" : "xhigh";
        if (difficulty >= 3.2 && Efforts.Rank(e) < Efforts.Rank("high")) e = "high";
        if (reasoning >= 2.75 && stakes >= 0.7 && difficulty >= 3.5) e = "max";
        return e;
    }

    /// <summary>Snap a desired effort to the nearest level the model supports, preferring the higher neighbour.</summary>
    public static string? SnapEffort(string want, IReadOnlyList<string> supported)
    {
        if (supported.Count == 0) return null;
        if (supported.Contains(want)) return want;
        var w = Efforts.Rank(want);
        var sorted = supported.OrderBy(Efforts.Rank).ToList();
        return sorted.FirstOrDefault(e => Efforts.Rank(e) >= w) ?? sorted[^1];
    }

    public static int EffectiveTier(ModelSpec m, string task)
    {
        var t = m.Tier;
        if (m.Strengths?.Contains(task) == true) t++;
        if (m.Weaknesses?.Contains(task) == true) t--;
        return t;
    }

    public static double EstimateCost(ModelSpec m, int inTokens, int outTokens, string? effort)
    {
        var mult = effort is not null && EffortMultiplier.TryGetValue(effort, out var x) ? x : 1;
        return (inTokens * m.Price.Input + outTokens * mult * m.Price.Output) / 1e6;
    }

    public static int ExpectedOutputTokens(double outputSize, int? requestedMax)
    {
        var i = Math.Max(0, Math.Min(2, (int)Math.Floor(outputSize)));
        var frac = outputSize - i;
        var est = i == 2 ? OutputTokens[2] : OutputTokens[i] + frac * (OutputTokens[i + 1] - OutputTokens[i]);
        return (int)Math.Round(requestedMax is { } max ? Math.Min(est, max) : est);
    }

    private static bool MeetsRequirements(ModelSpec m, Features f, int inTokens, int outTokens)
    {
        if (f.HasImages && !m.Vision) return false;
        if (f.HasTools && !m.Tools) return false;
        if (inTokens * 1.1 + outTokens > m.Context) return false;
        if (f.RequestedMaxTokens is { } req && req > m.MaxOutput) return false;
        return true;
    }

    private static ModelSpec? PickBaseline(IReadOnlyList<ModelSpec> available, string? id)
    {
        if (!string.IsNullOrEmpty(id))
        {
            var m = available.FirstOrDefault(x => x.Id == id || x.Model == id);
            if (m is not null) return m;
        }
        return available.OrderByDescending(m => m.Tier).ThenByDescending(m => m.Price.Output).FirstOrDefault();
    }
}

File: ts/src/policy.ts
import type { Candidate, Decision, Effort, Features, Judgment, ModelSpec, Task, Tier } from "./types.js";
import { EFFORT_LADDER } from "./types.js";

/**
 * Pure policy: judgments + features + catalog -> decision.
 * No I/O, so it is cheap to test and cheap to re-run with new weights.
 */
export interface PolicyOptions {
  /** Model the client would have used without frugal; defaults to the priciest frontier model. */
  baselineId?: string;
  /** Weights for the capability-need score. */
  weights?: { difficulty: number; reasoning: number; stakes: number };
  /** Need thresholds for tier 2 and tier 3. */
  tierThresholds?: { standard: number; frontier: number };
  /** Below this difficulty confidence, bump one tier up for safety. */
  minConfidence?: number;
  /** Stakes probability above which economy models are excluded. */
  highStakes?: number;
}

const DEFAULTS: Required<PolicyOptions> = {
  baselineId: "",
  weights: { difficulty: 0.55, reasoning: 0.3, stakes: 0.15 },
  tierThresholds: { standard: 0.28, frontier: 0.58 },
  minConfidence: 0.45,
  highStakes: 0.7,
};

/** Expected visible output tokens by output-size level (0, 1, 2). */
const OUTPUT_TOKENS = [250, 1200, 5000];
/** Output multiplier for reasoning tokens at each effort level. */
const EFFORT_MULTIPLIER: Record<Effort, number> = {
  none: 1, minimal: 1, low: 1.4, medium: 2.2, high: 3.5, xhigh: 5, max: 7,
};

export function decide(
  features: Features,
  judgment: Judgment,
  available: ModelSpec[],
  opts: PolicyOptions = {},
): Decision {
  const o = { ...DEFAULTS, ...opts };
  const rationale: string[] = [];
  const task = judgment.task.choice;
  const d = judgment.difficulty.score;
  const r = judgment.reasoning.score;
  const s = judgment.stakes;

  // 1. How much capability does this request need?
  const need = o.weights.difficulty * (d / 4) + o.weights.reasoning * (r / 3) + o.weights.stakes * s;
  let tier: Tier = need < o.tierThresholds.standard ? 1 : need < o.tierThresholds.frontier ? 2 : 3;
  rationale.push(
    `task=${task} difficulty=${d.toFixed(2)} reasoning=${r.toFixed(2)} stakes=${s.toFixed(2)} need=${need.toFixed(2)} -> tier ${tier}`,
  );

  if (judgment.source === "fallback") {
    rationale.push(`jev unavailable (${judgment.error ?? "unknown"}); using conservative defaults`);
  }
  if (s >= o.highStakes && tier === 1) {
    tier = 2;
    rationale.push(`stakes ${s.toFixed(2)} >= ${o.highStakes}: economy tier excluded`);
  }
  if (judgment.source === "jev" && judgment.difficulty.confidence < o.minConfidence && tier < 3) {
    tier = (tier + 1) as Tier;
    rationale.push(
      `difficulty confidence ${judgment.difficulty.confidence.toFixed(2)} < ${o.minConfidence}: bumped to tier ${tier}`,
    );
  }

  // 2. How hard should the chosen model think?
  const wantEffort = desiredEffort(d, r, s);
  rationale.push(`desired effort ${wantEffort}`);

  // 3. Filter candidates on hard requirements, then rank by expected cost.
  const outTokens = expectedOutputTokens(judgment.outputSize.score, features.requestedMaxTokens);
  const inTokens = features.inputTokens;
  const eligible = available.filter((m) => meetsRequirements(m, features, inTokens, outTokens));
  const scored: Candidate[] = eligible.map((m) => ({
    id: m.id,
    effectiveTier: effectiveTier(m, task),
    estimatedCost: estimateCost(m, inTokens, outTokens, snapEffort(wantEffort, m.effort)),
  }));

  let pool = scored.filter((c) => c.effectiveTier >= tier);
  if (pool.length === 0) {
    // Nothing at the required tier: take the most capable thing that fits.
    const best = Math.max(...scored.map((c) => c.effectiveTier), 0);
    pool = scored.filter((c) => c.effectiveTier === best);
    rationale.push(`no eligible model at tier ${tier}; using best available (tier ${best})`);
  }
  if (pool.length === 0) {
    throw new Error("frugal: no model satisfies the request (check provider keys, vision/tool support, context size)");
  }

  pool.sort((a, b) => a.estimatedCost - b.estimatedCost);
  const chosen = available.find((m) => m.id === pool[0].id)!;
  const effort = snapEffort(wantEffort, chosen.effort);
  if (effectiveTier(chosen, task) > chosen.tier) rationale.push(`${chosen.id} counts a tier higher for ${task}`);
  rationale.push(`cheapest eligible: ${chosen.id}${effort ? ` @ ${effort}` : ""} (~$${pool[0].estimatedCost.toFixed(5)})`);

  // 4. What would the baseline have cost?
  const baseline = pickBaseline(available, o.baselineId);
  const baselineCost = baseline
    ? estimateCost(baseline, inTokens, outTokens, snapEffort("high", baseline.effort))
    : pool[0].estimatedCost;

  return {
    model: chosen,
    effort,
    tier,
    task,
    judgment,
    estimate: {
      inputTokens: inTokens,
      outputTokens: outTokens,
      cost: pool[0].estimatedCost,
      baselineModel: baseline?.id ?? chosen.id,
      baselineCost,
      savings: Math.max(0, baselineCost - pool[0].estimatedCost),
    },
    baseline,
    rationale,
    candidates: scored.sort((a, b) => a.estimatedCost - b.estimatedCost),
    passthrough: false,
  };
}

export function desiredEffort(difficulty: number, reasoning: number, stakes: number): Effort {
  let e: Effort =
    reasoning < 0.5 ? "minimal" : reasoning < 1.25 ? "low" : reasoning < 2.0 ? "medium" : reasoning < 2.6 ? "high" : "xhigh";
  if (difficulty >= 3.2 && rank(e) < rank("high")) e = "high";
  if (reasoning >= 2.75 && stakes >= 0.7 && difficulty >= 3.5) e = "max";
  return e;
}

/** Snap a desired effort to the nearest level the model supports, preferring the higher neighbour. */
export function snapEffort(want: Effort, supported: Effort[]): Effort | undefined {
  if (!supported.length) return undefined;
  if (supported.includes(want)) return want;
  const w = rank(want);
  const sorted = [...supported].sort((a, b) => rank(a) - rank(b));
  return sorted.find((e) => rank(e) >= w) ?? sorted[sorted.length - 1];
}

export function rank(e: Effort): number {
  return EFFORT_LADDER.indexOf(e);
}

export function effectiveTier(m: ModelSpec, task: Task): number {
  let t: number = m.tier;
  if (m.strengths?.includes(task)) t += 1;
  if (m.weaknesses?.includes(task)) t -= 1;
  return t;
}

export function estimateCost(m: ModelSpec, inTokens: number, outTokens: number, effort?: Effort): number {
  const mult = effort ? EFFORT_MULTIPLIER[effort] : 1;
  return (inTokens * m.price.input + outTokens * mult * m.price.output) / 1e6;
}

export function expectedOutputTokens(outputSize: number, requestedMax?: number): number {
  const i = Math.max(0, Math.min(2, Math.floor(outputSize)));
  const frac = outputSize - i;
  const est = i === 2 ? OUTPUT_TOKENS[2] : OUTPUT_TOKENS[i] + frac * (OUTPUT_TOKENS[i + 1] - OUTPUT_TOKENS[i]);
  return Math.round(requestedMax ? Math.min(est, requestedMax) : est);
}

function meetsRequirements(m: ModelSpec, f: Features, inTokens: number, outTokens: number): boolean {
  if (f.hasImages && !m.vision) return false;
  if (f.hasTools && !m.tools) return false;
  if (inTokens * 1.1 + outTokens > m.context) return false;
  if (f.requestedMaxTokens && f.requestedMaxTokens > m.maxOutput) return false;
  return true;
}

function pickBaseline(available: ModelSpec[], id?: string): ModelSpec | undefined {
  if (id) {
    const m = available.find((x) => x.id === id || x.model === id);
    if (m) return m;
  }
  return [...available].sort((a, b) => b.tier - a.tier || b.price.output - a.price.output)[0];
}
